core/src/Streamly/Internal/Data/Fold.hs

```diff
@@ -1339,21 +1344,160 @@ takeEndBy predicate (Fold fstep finitial fextract) =
 -- Binary splitting on a separator
 ------------------------------------------------------------------------------
 
-{-
--- | Find the first occurrence of the specified sequence in the input stream
--- and break the input stream into two parts, the first part consisting of the
--- stream before the sequence and the second part consisting of the sequence
--- and the rest of the stream.
+data SplitOnSeqState acc a rb rh w ck =
+      SplitOnSeqEmpty !acc
+    | SplitOnSeqSingle !acc !a
+    | SplitOnSeqWord !acc !Int !w
+    | SplitOnSeqWordLoop !acc !w
+    | SplitOnSeqKR !acc !Int !rb !rh
+    | SplitOnSeqKRLoop !acc !ck !rb !rh
+
+-- XXX Need to add tests for takeEndBySeq, we have tests for takeEndBySeq_ .
+
+-- | Continue taking the input until the input sequence matches the supplied
+-- sequence, taking the supplied sequence as well. If the pattern is empty this
+-- acts as an identity fold.
 --
--- > let breakOn_ pat xs = S.fold (S.breakOn pat FL.toList FL.toList) $ S.fromList xs
+-- >>> s = Stream.fromList "hello there. How are you?"
+-- >>> f = Fold.takeEndBySeq (Array.fromList "re") Fold.toList
+-- >>> Stream.fold f s
+-- "hello there"
 --
--- >>> breakOn_ "dear" "Hello dear world!"
--- > ("Hello ","dear world!")
+-- >>> Stream.toList $ Stream.foldMany f s
+-- ["hello there", ". How are", "you?"]
 --
-{-# INLINE breakOn #-}
-breakOn :: Monad m => Array a -> Fold m a b -> Fold m a c -> Fold m a (b,c)
-breakOn pat f m = undefined
--}
+-- /Pre-release/
+{-# INLINE takeEndBySeq #-}
+takeEndBySeq :: forall m a b. (MonadIO m, Storable a, Enum a, Eq a) =>
+       Array.Array a
+    -> Fold m a b
+    -> Fold m a b
+takeEndBySeq patArr (Fold fstep finitial fextract) =
```

Comment:
You could probably implement `takeEndBySeq` and `takeEndBySeq_` in the same function if you let it take another argument, `withSep`, like we do it in `splitOnSuffixSeq`

Reply:
had thought about that but step is significantly different and extract is entirely different, so it would make the code complicated to understand, and I was not sure how perf might get impacted. You can try it out.